src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
+        // which is exactly what we want (see above). Additionally...
+        //
+        // The old Win32 clipboard API as used below is somewhere in the order of 1000x faster than the WinRT one,
+        // which has about the same latency as a fibre optics connection across the entire WA state.
+        // Don't use the WinRT clipboard API. RPC calls are not funny.
+        winrt::hstring text;
+        {
+            // According to various reports on the internet, OpenClipboard might
+            // fail to acquire the internal lock over RDP, due to rdpclip.exe.
+            // I haven't seen this happen yet myself however.
+            for (int attempts = 0;;)
             {
-                auto items = co_await data.GetStorageItemsAsync();
-                if (items.Size() > 0)
+                if (!OpenClipboard(nullptr))
                 {
-                    auto item = items.GetAt(0);
-                    text = item.Path();
+                    if (++attempts > 5)
+                    {
+                        co_return;
+                    }
+
+                    Sleep(5);
+                    continue;
                 }
+
+                break;
             }
 
-            if (_settings.GlobalSettings().TrimPaste())
+            const auto clipboardCleanup = wil::scope_exit([]() {
+                CloseClipboard();
+            });
+
+            const auto data = GetClipboardData(CF_UNICODETEXT);
+            if (!data)
             {
-                text = { Utils::TrimPaste(text) };
-                if (text.empty())
-                {
-                    // Text is all white space, nothing to paste
-                    co_return;
-                }
+                co_return;
             }
 
-            // If the requesting terminal is in bracketed paste mode, then we don't need to warn about a multi-line paste.
-            auto warnMultiLine = _settings.GlobalSettings().WarnAboutMultiLinePaste() &&
-                                 !eventArgs.BracketedPasteEnabled();
-            if (warnMultiLine)
+            const auto str = static_cast<const wchar_t*>(GlobalLock(data));
+            if (!str)
             {
-                const auto isNewLineLambda = [](auto c) { return c == L'\n' || c == L'\r'; };
-                const auto hasNewLine = std::find_if(text.cbegin(), text.cend(), isNewLineLambda) != text.cend();
-                warnMultiLine = hasNewLine;
+                co_return;
             }
 
-            constexpr const std::size_t minimumSizeForWarning = 1024 * 5; // 5 KiB
-            const auto warnLargeText = text.size() > minimumSizeForWarning &&
-                                       _settings.GlobalSettings().WarnAboutLargePaste();
+            const auto dataCleanup = wil::scope_exit([&]() {
+                GlobalUnlock(data);
+            });
+
+            text = winrt::hstring{ str, gsl::narrow_cast<uint32_t>(wcslen(str)) };
```

Comment:
I feel like the new code is very verbose, but also very easy to follow. If it's too verbose for you I wouldn't mind abstracting away the `OpenClipboard` / `CloseClipboard` calls into a safer wrapper like `wil::open_clipboard` (which lacks the retrying logic) and something like `wil::unique_hglobal_locked` (which unfortunately is `void*` and not `wchar_t*`).

IMO it's fine this way because this is the only use of this code and I wouldn't introduce an abstraction until we got 2-3 uses of this (at which point we'll probably have a better idea how the data types should look like, especially in respect to `SetClipboardData`).